File: 2015 ICPC North American Qualifier Contest/countcircuits.cs
using System;
using System.Collections.Generic;

class Program {
    static void Generate(List<(int, int)> v, Dictionary<(int, int), int> map) {
        int n = v.Count;
        for (int i = 1; i < (1 << n); i++) {
            (int, int) sum = (0, 0);
            for (int j = 0; j < n; j++) {
                if (((1 << j) & i) != 0) {
                    sum.Item1 += v[j].Item1;
                    sum.Item2 += v[j].Item2;
                }
            }
            if (map.ContainsKey(sum)) {
                map[sum]++;
            } else {
                map[sum] = 1;
            }
        }
    }

    static void Main() {
        int n = int.Parse(Console.ReadLine());
        
        var v1 = new List<(int, int)>();
        var v2 = new List<(int, int)>();
        
        for (int i = 0; i < n; i++) {
            string[] parts = Console.ReadLine().Split();
            int first = int.Parse(parts[0]);
            int second = int.Parse(parts[1]);
            
            if (i % 2 == 0) {
                v1.Add((first, second));
            } else {
                v2.Add((first, second));
            }
        }
        
        var set1 = new Dictionary<(int, int), int>();
        var set2 = new Dictionary<(int, int), int>();
        
        Generate(v1, set1);
        Generate(v2, set2);
        
        long ans = 0;
        
        foreach (var pair in set1) {
            var p = pair.Key;
            var target = (-p.Item1, -p.Item2);
            if (set2.ContainsKey(target)) {
                ans += pair.Value * set2[target];
            }
        }
        
        // Add empty subsets
        var zero = (0, 0);
        if (set1.ContainsKey(zero)) ans += set1[zero];
        if (set2.ContainsKey(zero)) ans += set2[zero];
        
        Console.WriteLine(ans);
    }
}

File: 2015 ICPC North American Qualifier Contest/countcircuits.java
import java.util.*;
import java.io.*;

public class countcircuits {
    static class Pair {
        int first, second;
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
        
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Pair pair = (Pair) o;
            return first == pair.first && second == pair.second;
        }
        
        @Override
        public int hashCode() {
            return Objects.hash(first, second);
        }
    }
    
    static void generate(List<Pair> v, Map<Pair, Integer> map) {
        int n = v.size();
        for (int i = 1; i < (1 << n); i++) {
            Pair sum = new Pair(0, 0);
            for (int j = 0; j < n; j++) {
                if (((1 << j) & i) != 0) {
                    sum.first += v.get(j).first;
                    sum.second += v.get(j).second;
                }
            }
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        
        List<Pair> v1 = new ArrayList<>();
        List<Pair> v2 = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            int first = scanner.nextInt();
            int second = scanner.nextInt();
            if (i % 2 == 0) {
                v1.add(new Pair(first, second));
            } else {
                v2.add(new Pair(first, second));
            }
        }
        
        Map<Pair, Integer> set1 = new HashMap<>();
        Map<Pair, Integer> set2 = new HashMap<>();
        
        generate(v1, set1);
        generate(v2, set2);
        
        long ans = 0;
        
        for (Map.Entry<Pair, Integer> entry : set1.entrySet()) {
            Pair p = entry.getKey();
            Pair target = new Pair(-p.first, -p.second);
            if (set2.containsKey(target)) {
                ans += entry.getValue() * set2.get(target);
            }
        }
        
        // Add empty subsets
        Pair zero = new Pair(0, 0);
        if (set1.containsKey(zero)) ans += set1.get(zero);
        if (set2.containsKey(zero)) ans += set2.get(zero);
        
        System.out.println(ans);
        
        scanner.close();
    }
}